kv_connectors/llmd_fs_backend/src/csrc/storage/storage_offload.cu

```diff
@@ -0,0 +1,362 @@
+            bool stage1_ok = false;
+            torch::Tensor host_buf;
+            try {
+                // Read data from disk into pinned memory tensor.
+                host_buf = TIME_EXPR("read phase 1: read_file_from_disk", read_file_from_disk(src_file), ("file:" + src_file));
+                stage1_ok = true;
+            } catch (const std::exception& e) {
+                std::cerr << "[ERROR] Stage1 read_file_from_disk failed for " << src_file << ": " << e.what() << std::endl;
+            } catch (...) {
+                std::cerr << "[ERROR] Stage1 unknown failure for " << src_file << std::endl;
+            }
+
+            // Stage 2:  copy tensors from pinned CPU buffer to GPU.
+            bool ok = false;
+            if (stage1_ok) {
+                try {
+                    // Perform asynchronous GPU copy and tensor swap.
+                    ok = TIME_EXPR("read phase 2: copy_buffer_to_gpu_tensors",
+                                   copy_buffer_to_gpu_tensors(host_buf, block_ids, dst_tensors, gpu_blocks_per_file, *thread_stream),
+                                   "file: " + src_file);
+                    cudaError_t err = cudaStreamSynchronize(thread_stream->stream());
+                    if (err != cudaSuccess) {
+                        std::cerr << "[ERROR] cudaStreamSynchronize failed: " << cudaGetErrorString(err) << std::endl;
+                        ok = false;
+                    }
+                } catch (const std::exception& e) {
+                    std::cerr << "[ERROR] Stage2 copy_and_swap failed for " << src_file << ": " << e.what() << std::endl;
+                } catch (...) {
+                    std::cerr << "[ERROR] Stage2 unknown failure for " << src_file << std::endl;
+                }
+            }
+
+            // Final cleanup & accounting
+            // Synchronize only this thread's CUDA stream.
+            at::cuda::setCurrentCUDAStream(current_stream);
+            job_state->completed_tasks.fetch_add(1);
+            if (!ok) job_state->all_success = false;
+            return ok;
+        });
+
+        // Convert std::future → std::shared_future- is copyable and can be waited on by multiple threads.
+        job_state->futures.push_back(future.share());
+    }
+
+    std::lock_guard<std::mutex> lock(jobs_mutex);
+    jobs[job_id] = std::move(job_state);
+    return true;
+}
+
+// -------------------------------
+// PYBIND11 module
+// -------------------------------
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+    m.def("init_resources",
+          &init_resources,
+          py::arg("io_threads"),
+          py::arg("pinned_buffer_size_mb"),
+          py::arg("max_pinned_memory_gb"),
+          py::arg("tp_rank"),
+          py::arg("kv_before_blocks"),
+          py::arg("layers_before_blocks"),
+          py::arg("block_axis"));
+
+    m.def("cleanup_resources", &cleanup_resources);
+
+    m.def("get_finished", &get_finished);
+
+    m.def("transfer_async_put",
```

Comment:
nit:
I think `async_save_gpu_blocks` and `async_load_gpu_blocks` are clearer names for `transfer_async_put` and `transfer_async_get`.